core/src/repair_weighted_traversal.rs

```diff
@@ -69,10 +69,119 @@ impl<'a> Iterator for RepairWeightTraversal<'a> {
     }
 }
 
+struct GenericTraversal<'a> {
+    tree: &'a HeaviestSubtreeForkChoice,
+    pending: Vec<Slot>,
+}
+
+impl<'a> GenericTraversal<'a> {
+    pub fn new(tree: &'a HeaviestSubtreeForkChoice) -> Self {
+        Self {
+            tree,
+            pending: vec![tree.root().0],
+        }
+    }
+}
+
+impl<'a> Iterator for GenericTraversal<'a> {
+    type Item = Slot;
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.pending.pop();
+        if let Some(slot) = next {
+            let children: Vec<_> = self
+                .tree
+                .children(&(slot, Hash::default()))
+                .unwrap()
+                .iter()
+                .map(|(child_slot, _)| *child_slot)
+                .collect();
+            self.pending.extend(children);
+        }
+        next
+    }
+}
+
+pub fn get_unknown_last_index(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut repairs = Vec::new();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if repairs.len() >= limit {
+            break;
+        }
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.known_last_index().is_none() {
+                repairs.push(ShredRepairType::HighestShred(slot, slot_meta.received));
+                processed_slots.insert(slot);
+            }
+        }
+    }
+    repairs
+}
+
+pub fn get_closest_completion(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut v: Vec<(Slot, u64)> = Vec::default();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.is_full() {
+                continue;
+            }
+            if let Some(last_index) = slot_meta.known_last_index() {
+                let dist = last_index - slot_meta.consumed;
+                v.push((slot, dist));
+                processed_slots.insert(slot);
+            }
+        }
+    }
+    v.sort_by(|(_, d1), (_, d2)| d1.cmp(d2));
```

Comment:
Hmmm I'm wondering if we should avoid repairing a slot if it's really close to completion and instead prioritize it's parents. This is because we could have a situation like:

4 (far from completion) -> 5 (far from completion) -> 6 (far from completion) -> 8 (really close to completion)

We could repair `8`, but we won't be able to replay anything until all it's parents are repaired

Reply:
This actually brings me to this scenario, perhaps outside of the scope of this PR. Imagine we had a network partition such that we had a few different forks, looks like


```
   /- 1 - 2 - 3 - 4 (33%) ------|                                       |------- 15 - 16
0  - 5 - 6- 7 (33%)----------| PARTITION RESOLVES| -------- 20 - 21
   \ - 9 - 10 - 11 - 12 (33%)-- |                                       | --------- 30 - 31
```

Everything before the partition has very low completion rate, everything after has really high completion rate. In this case, just repairing based on completion rate might be insufficient, we might need to rely on leader retransmit.

I'm thinking in this case maybe we retransmit as long as you believe your fork is the heaviest fork based on the `heaviest_subtree_fork_choice` in replay

Reply:
Updated this to sort slots by completion, then for each slot prioritize repairing incomplete parent slots in order.

I'm looking at retransmit as a separate PR, so I'd prefer to limit this PR to just these repairs.